### ISSUE TITLE
Unused Method

### ISSUE THREAD

**Issue by JoeGaebel** [arturbosch/detekt#2229]
## Expected Behavior

```kotlin
fun unusedMethod() {
    println("cool!")
}
```

Running detekt, I'd love a rule that caught unused methods.
## Current Behavior

Currently there is only Unused Private Methods as a rule.

## Context

Unused public functions, in my opinion, are just as bad as unused private functions - it's dead code, and I suspect this would be a useful rule.

**Comment by imanushin** [comment]
There are some corner cases, which should be also covered:
* ```override``` methods should be marked as "used"
* ```internal``` methods (or from ```internal```/```private``` classes) can be checked just by using current project details
* ```public``` methods, which can be used by another modules, can be checked, however this requires special trick: ```detekt``` should pre-compile all other modules to validate this. E.g. if you project has subprojects with dependencies A --> B (e.g. A references B) then task ```B.detect()``` should wait for compilation of project A
* This rule should be configurable to exclude methods, which are accessible via reflection. For example, [Spring IoC can call methods via xml configuration](https://stackoverflow.com/questions/5419695/init-method-in-spring-controller-annotation-version). So looks like we have to put optional list of ignored annotations into this rule configuration.

**Comment by schalkms** [comment]
Thanks for your feedback!

This behavior is not so easy to implement, since there are many many edge cases. The mentioned rule needs certainly type and symbol solving in order to properly work. This rule will also benefit from type and symbol solving if we don't add any new behavior. That's why I added to our list in issue #1915. I don't want to have two different issues for basically the same thing. Therefore, I'll close this issue but keep the discussion open.

**Comment by arturbosch** [comment]
@imanushin thanks for the list!

I've experimented with dead code detection and it's clearly possible with the compiler plugin architecture to come up with a good rule.

**Comment by imanushin** [comment]
@arturbosch , much appreciated of your effort.

**Comment by Sevastyan** [comment]
So, is there a rule to detect dead public code, or not?

**Comment by schalkms** [comment]
> So, is there a rule to detect dead public code, or not?

Not as part of detekt.
For 3rd party rulesets I can't say.
